server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletMutationPrepAttempt.java

```diff
@@ -0,0 +1,124 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.tserver.tablet;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.accumulo.core.constraints.Violations;
+import org.apache.accumulo.core.data.Mutation;
+
+public class TabletMutationPrepAttempt {
+
+  private final Violations violations = new Violations();
+
+  private CommitSession commitSession;
+  private List<Mutation> violators;
+  private List<Mutation> nonViolators;
+  private boolean attemptedTabletPrep;
+
+  /**
+   * Return true if an attempt was made to update the writes in progress for the tablet and to
+   * retrieve the commit session from the tablet's memory, or false otherwise.
+   */
+  public boolean attemptedTabletPrep() {
+    return attemptedTabletPrep;
+  }
+
+  /**
+   * Return true if this attempt has a non-null commit session. If null, it indicates that either
+   * there was never an attempt made to get the commit session from the tablet's memory, or that
+   * either the tablet has a closed state or the tablet's memory is unavailable.
+   */
+  public boolean hasCommitSession() {
+    return commitSession != null;
+  }
+
+  /**
+   * Retrieve the commit session. Possibly null.
+   */
+  public CommitSession getCommitSession() {
+    return commitSession;
+  }
+
+  /**
+   * Retrieve the constraint violations found across the mutations. Never null.
+   */
+  public Violations getViolations() {
+    return violations;
+  }
+
+  /**
+   * Return true if at least one constraint violation was found, or false otherwise.
+   */
+  public boolean hasViolations() {
+    return !violations.isEmpty();
+  }
+
+  /**
+   * Return the list of mutations that violated a constraint. Either null or a non-empty list.
+   */
+  public List<Mutation> getViolators() {
+    return violators;
+  }
+
+  /**
+   * Return the list of mutations that did not violate any constraints. Either null or a non-empty
+   * list.
+   */
+  public List<Mutation> getNonViolators() {
+    return nonViolators;
+  }
+
+  /**
+   * Return true if at least one mutation that did not violate any constraints was found.
+   */
+  public boolean hasNonViolators() {
+    return nonViolators != null && !nonViolators.isEmpty();
+  }
+
+  /**
+   * Set the commit session that was retrieved from the tablet's memory, and set
+   * {@link #attemptedTabletPrep} to true.
+   */
+  void setCommitSession(final CommitSession commitSession) {
+    this.commitSession = commitSession;
+    attemptedTabletPrep = true;
+  }
+
+  /**
+   * Add a mutation that violates a constraint, along with information about the constraints that it
+   * violated.
+   */
+  void addViolator(final Mutation mutation, final Violations mutationViolations) {
+    if (violators == null) {
+      violators = new ArrayList<>();
```

Comment:
Not thread-safe in a way that could destroy state. Have you made sure that this is only ever used in a single thread? This could be related to my earlier comment about null-ness consistency. You could make sure that this member variable is initialized when this object is constructed, as long as "null" vs. "empty" isn't a meaningful distinction.

Reply:
I'll modify it so that the lists are initialized when the object is constructed. From my understanding of the original code, null vs. not is only a meaningful distinction for the `CommitSession` variable.